### PR TITLE
Validation exception for max/min filter function on non numeric fields.

### PR DESCRIPTION
**Pull request by sanjanasandeep**
**Issue**

https://cdap.atlassian.net/browse/CDAP-18919

Addressing comments from this closed PR: https://github.com/cdapio/hydrator-plugins/pull/1551

### CODE REVIEW

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregator.java [hunk 1]
```diff
@@ -186,19 +187,35 @@ private void validateSchema(Schema inputSchema, List<String> uniqueFields,
...
+                        null)
+                .withConfigProperty("filterOperation");
       }
-    }
+      List<Schema.Type> numericSchemaTypes = new ArrayList(Arrays.asList(Schema.Type.INT, Schema.Type.LONG,
+              Schema.Type.FLOAT, Schema.Type.DOUBLE));
+      Schema.Type schemaType;
+      if ((field.getSchema().getType()).equals(Schema.Type.UNION)) {
+        schemaType = field.getSchema().getUnionSchema(0).getType();
+        ;
```
Comment (fernst) [inline core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregator.java]: extra semicolon

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregator.java [hunk 2]
```diff
@@ -186,19 +187,35 @@ private void validateSchema(Schema inputSchema, List<String> uniqueFields,
...
-          .withConfigProperty("filterOperation");
+                                function.getFunction(), function.getField(), function.getField()),
+                        null)
+                .withConfigProperty("filterOperation");
       }
-    }
+      List<Schema.Type> numericSchemaTypes = new ArrayList(Arrays.asList(Schema.Type.INT, Schema.Type.LONG,
+              Schema.Type.FLOAT, Schema.Type.DOUBLE));
+      Schema.Type schemaType;
+      if ((field.getSchema().getType()).equals(Schema.Type.UNION)) {
```
Comment (fernst) [inline core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregator.java]: Are you checking Unions to check for nullable fields?

If so, here's the quickest way to do this:

```
Schema.Type fieldType = field.getSchema().isNullable() ? 
  field.getSchema().getNonNullable().getType() : field.getSchema().getType();
```

File: core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/DedupAggregatorValidationTest.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@RunWith(MockitoJUnitRunner.class)
+public class DedupAggregatorValidationTest {
```
Comment (fernst) [inline core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/DedupAggregatorValidationTest.java:35]: great test 👍

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregator.java (hunk 2 above)
Comment (sanjanasandeep) [reply core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregator.java]: got it. addressed this in the next rev

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregator.java [hunk 4]
```diff
@@ -186,19 +187,30 @@ private void validateSchema(Schema inputSchema, List<String> uniqueFields,
...
         collector.addFailure(String.format("Invalid filter %s(%s): Field '%s' does not exist in input schema ",
-                                           function.getFunction(), function.getField(), function.getField()),
-                             null)
-          .withConfigProperty("filterOperation");
+                                function.getFunction(), function.getField(), function.getField()),
+                        null)
+                .withConfigProperty("filterOperation");
       }
-    }
+      List<Schema.Type> numericSchemaTypes = new ArrayList(Arrays.asList(Schema.Type.INT, Schema.Type.LONG,
```
Comment (tivv) [inline core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregator.java]: Can it be an `EnumSet` constant? I don't think you need to build it each time

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregator.java [hunk 5]
```diff
@@ -186,19 +187,30 @@ private void validateSchema(Schema inputSchema, List<String> uniqueFields,
...
+              Schema.Type.FLOAT, Schema.Type.DOUBLE));
+      Schema.Type fieldType = field.getSchema().isNullable() ?
+              field.getSchema().getNonNullable().getType() : field.getSchema().getType();
+      if ((function.getFunction().equals(DedupConfig.Function.MAX) ||
+              function.getFunction().equals(DedupConfig.Function.MIN)) && !numericSchemaTypes.contains(fieldType)) {
+        collector.addFailure(String.format("Unsupported filter operation %s(%s): Field '%s' is non numeric ",
```
Comment (fernst) [inline core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregator.java:208]: nit: arrange this so the max or min check is in the same line for better legibility:

```
      if ((function.getFunction().equals(DedupConfig.Function.MAX) || function.getFunction().equals(DedupConfig.Function.MIN)) 
        && !numericSchemaTypes.contains(fieldType)) {
 ```

**Review by fernst — APPROVED** [review]

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregator.java [hunk 6]
```diff
@@ -186,19 +187,30 @@ private void validateSchema(Schema inputSchema, List<String> uniqueFields,
...
+                                function.getFunction(), function.getField(), function.getField()),
+                        null)
+                .withConfigProperty("filterOperation");
       }
-    }
+      List<Schema.Type> numericSchemaTypes = new ArrayList(Arrays.asList(Schema.Type.INT, Schema.Type.LONG,
+              Schema.Type.FLOAT, Schema.Type.DOUBLE));
+      Schema.Type fieldType = field.getSchema().isNullable() ?
+              field.getSchema().getNonNullable().getType() : field.getSchema().getType();
+      if ((function.getFunction().equals(DedupConfig.Function.MAX) ||
```
Comment (tivv) [inline core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregator.java]: You don't need equals for enums

File: core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/DedupAggregatorValidationTest.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+    @Before
+    public void setUp() {
+        collector = new MockFailureCollector();
+        Schema stringSchema = Schema.of(Schema.Type.STRING);
+        Schema intSchema = Schema.of(Schema.Type.INT);
+        Schema inputSchema = Schema.recordOf(
+                Schema.Field.of("name", Schema.nullableOf(stringSchema)),
+                Schema.Field.of("profession", Schema.nullableOf(stringSchema)),
+                Schema.Field.of("age", Schema.nullableOf(intSchema))
```
Comment (tivv) [inline core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/DedupAggregatorValidationTest.java:57]: Could you please test on decimal field

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregator.java (hunk 6 above)
Comment (fernst) [reply core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregator.java]: ```function.getFunction() == DedupConfig.Function.MAX``` is equivalent

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregator.java (hunk 4 above)
Comment (sanjanasandeep) [reply core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregator.java]: yeah thats a good point.

File: core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregator.java (hunk 6 above)
Comment (sanjanasandeep) [reply core-plugins/src/main/java/io/cdap/plugin/batch/aggregator/DedupAggregator.java]: yup also `==` is null safe (https://stackoverflow.com/questions/1750435/comparing-java-enum-members-or-equals)

File: core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/DedupAggregatorValidationTest.java (hunk 7 above)
Comment (sanjanasandeep) [reply core-plugins/src/test/java/io/cdap/plugin/batch/aggregator/DedupAggregatorValidationTest.java:57]: as discussed offline, I think we address this in the next PR. Currently it is not supported.

**Review by fernst — APPROVED** [review]
LGTM

File: core-plugins/docs/Deduplicate-batchaggregator.md [hunk 8]
```diff
@@ -4,7 +4,8 @@
 Description
 -----------
 De-duplicates records either using one or more fields or by using the record as a whole. Additionally, it supports logically
-choosing a record out of the duplicate records based on a filter field. Supported logical functions are `any`, `max` and `min`. 
+choosing a record out of the duplicate records based on a filter field. Supported logical functions are `any`, `max` and `min`.
+`min` and `max` must be operated only on numeric filter fields.
```
Comment (rrielley76) [inline core-plugins/docs/Deduplicate-batchaggregator.md:8]: @sanjanasandeep Do you need to add this sentence to the Filter property description too?

Comment (sanjanasandeep) [reply core-plugins/docs/Deduplicate-batchaggregator.md:8]: yes, that's a good point. That would need to be updated too

**Pull request merged**